proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+
+ * They could be defined in a new `Cabal-hooks` library, with `Cabal` depending
+   on it and reusing the same types. This has the advantage of more clearly
+   defining the subset of the `Cabal` library that is exposed via the hooks API,
+   which is helpful for modularity. However it is likely to involve moving quite
+   a lot of code to the new library.
+
+ * Separate hook input/output data types could be defined in a `Cabal-hooks`
+   library, then `Cabal` could contain code for translating values of those
+   datatypes into its internal representations (and vice versa).  This would
+   require significant duplication of datatypes, but might provide more options
+   for evolving the `Cabal` and `Cabal-hooks` datatypes independently.
+
+A related question is whether the hooks API should be versioned separately from
+`Cabal` itself.  Version compatibility concerns may arise because the build tool
+and `SetupHooks.hs` may use different versions of the `Cabal` library. The
+impact depends on how the build tool is using `SetupHooks.hs`:
+
+ - If the build tool is compiling a shim `Setup.hs`, it can in principle pick a
+   compatible version of `Cabal` even if the build tool itself uses a different
+   version, provided various constraints are satisfied (both those from the
+   `setup-depends` of the package itself, and those arising from the build tool,
+   e.g. [`cabal-install`](https://github.com/haskell/cabal/blob/c97092f5af484dfd5c1a465e0754114571264447/cabal-install/src/Distribution/Client/ProjectPlanning.hs#L1419-L1454)).
+
+ - If the build tool is serialising data structures from its version of
+   `Cabal[-hooks]` to the version against which the "hooks executable" including
+   `SetupHooks.hs` is built, then the serialisation formats need to be
+   compatible.  We could imagine using a serialisation format that is versioned
+   and has some support for migrations (cf. `safecopy`) but this would lead to
+   additional complexity.
+
+ - If the build tool is dynamically loading `SetupHooks.hs`, ABI compatibility
+   means both will need to use exactly the same `Cabal[-hooks]` library.
+
+It is important to be clear what future compatibility guarantees (if any) are
+offered by `Cabal` and/or build tools to packages using `Hooks`.  There is a
+tension here, because we do not want package maintainers to be over-burdened
+with continual changes to support newer `Cabal` versions, but neither do we want
+`Cabal` maintainers to be over-burdened by the costs of providing backwards
+compatibility.
+
+Since `Hooks` allows the build system to be extended, arguably packages using
+this feature necessarily depend on the version of the build system (and hence
+the `Cabal` version). Thus it may be reasonable to expect package authors to
+declare the range of `Cabal` versions for which their package works (in
+`setup-depends`) and require it to be updated to support newer versions of build
+tools.
```

Comment:
I think we should version the hooks interface separate than the Cabal implementation.

Otherwise we'll just have more "bumping the upper bound" busywork, which has proved to be extremely toxic community wise (people just go rogue and have bad upper bounds to avoid chores).

Reply:
We currently require the Cabal version the Setup works against in setup-depends, and its fine. This is because that version does _not_ have to correspond to the Cabal version that `cabal-install` is linked against, and we ensure older ghcs can compile newer `Cabal` libraries, and conversely newer ghcs are typically able to install much older `Cabal` libraries. So this would just be keeping with established practice in one of the situations where that practice seems to have been working fine.

Reply:
@gbaz I totally agree with that the library used in the host `cabal-install` can be different. I just think we can do better than a `Cabal` version bound:

The whole point of this change is the interface is now much more limited than "I will do whatever I like with the right CLI, and oh btw I am depending on the `Cabal` library and might used anything part of it"

If we depend on `Cabal` we are still depending on all sorts of code implementing functionality that is not supposed to be controlled by the custom hooks! Perhaps there is no reason to use that stuff from a hook, but there is no reason that it couldn't be --- the `Cabal` library is still massive overkill and belies any new invariants we are trying to enforce.

We should make a tiny little library which is as stripped down as much as possible, and purposely doesn't include in its closure any "implementation" stuff that the individual package is no longer supposed to be in charge of. That "proves" we have the intended separation of concerns.

`Cabal` and `cabal-install`, as needed, can depend on that tiny little library too, and of course they are free to depend upon a different version.

Reply:
I don't think this problem is coupled with enforcing invariants or the like -- that comes from the types in play. Hooks are free to import whatever other libraries they'd like, regardless. 

Because we can vary the `Cabal` library we depend on separate from other concerns, the argument for depending on a smaller independently versioned library comes down to aesthetics. My view is that as long as `LocalBuildInfo` (or even a slimmed-down refactor of it) is given as an input and output to various hooks, then the versioning is necessarily pretty tightly coupled to Cabal. For example, LBI contains `ConfigFlags`, which are very much coupled to the command line interface of Cabal, and don't even make sense to drop down to the cabal-syntax package. `LocalBuildInfo` also contains `ComponentLocalBuildInfo` and that's full of cabal specific things like UnitId, MungedPackageId, and soforth. If we were to go so far as to present some data structure that _omitted_ all this stuff, then I could possibly see decoupling. But even for my inclination to slim the API surface and make the types more granular, I think that's going too far for a refactor pass, and could likely cut off some desired use-cases.

Again -- if we got something resembling a real win out of doing this work, I could see the motivation. But it doesn't buy us any more invariants. Not only can users depend on the whole `Cabal` library regardless, but by design, they can depend on any other code from hackage they desire. And as I first argued, decoupling versioning doesn't buy us anything either.

Reply:
Users being allowed to depend on stuff != Users being forced to depend on stuff.

If the proposed hooks interface "morally" drags into all of `Cabal` (e.g. because of a lon tail of concepts in `LocalBuildInfo`) that to me is a clear sign that the proposed hooks interface is bad and doesn't meet the stated objectives.

I can be persuaded to accept this as a distasteful stop-gap just like having too many hooks, but the final interface we actually want to end up with *must* be lighter weight. This is crucially important to me — the same lesson as cutting out `DynFlags` and `HscEnv`.